src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperCollector.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+public abstract partial class TagHelperCollector<T>
+    where T : ITagHelperDescriptorProvider
+{
+    // This type is generic to ensure that each descendent gets its own instance of this field.
+    private static readonly ConditionalWeakTable<IAssemblySymbol, Cache> s_perAssemblyCaches = new();
+
+    private readonly Compilation _compilation;
+    private readonly ISymbol? _targetSymbol;
+
+    protected TagHelperCollector(Compilation compilation, ISymbol? targetSymbol)
+    {
+        _compilation = compilation;
+        _targetSymbol = targetSymbol;
+    }
+
+    private static bool MightContainTagHelpers(IAssemblySymbol assembly)
+    {
+        // In order to contain tag helpers, components, or anything else we might want to find,
+        // the assembly must start with "Microsoft.AspNetCore." or reference an assembly that
+        // starts with "Microsoft.AspNetCore."
+
+        return assembly.Name.StartsWith("Microsoft.AspNetCore.", StringComparison.Ordinal) ||
+               assembly.Modules.First().ReferencedAssemblies.Any(
+                   a => a.Name.StartsWith("Microsoft.AspNetCore.", StringComparison.Ordinal));
```

Comment:
Looks like this change closes https://github.com/dotnet/razor/issues/8336.

Reply:
Good catch! Yes, it does close that.